All remaining linear/element-wise formulas

albanD opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#59993 All remaining linear/element-wise formulas**
* #60499 Fix out= variant forward grad detection
* #60498 Add forward AD inplace check and fix codegen

Differential Revision: [D29914594](https://our.internmc.facebook.com/intern/diff/D29914594)

albanD committed:
All remaining linear/element-wise formulas

[ghstack-poisoned]

zou3519 commented on tools/autograd/derivatives.yaml:
```diff
@@ -1092,21 +1098,27 @@
 - name: remainder.Tensor(Tensor self, Tensor other) -> Tensor
   self: grad
+  # Use addition with zeros to get the proper broadcasting behavior
+  result: self_t + at::zeros_like(other)
```
Can you use result? `self_t.expand_as(result)`.

Otherwise, there are ways to compute the broadcast shape like infer_size: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/ExpandUtils.h#L14

albanD replied:
The problem with expand is that it needs to have the same number of dimensions.
But yes, let me check infer_size!

zou3519 replied:
> The problem with expand is that it needs to have the same number of dimensions.

Does it?
```
[ins] In [1]: import torch
x = torch.
[ins] In [2]: x = torch.randn(2, 1)

[ins] In [3]: y = torch.randn(3, 1, 4)

[ins] In [4]: result = x + y

[ins] In [5]: x.expand_as(result)
Out[5]:
tensor([[[ 0.7218,  0.7218,  0.7218,  0.7218],
         [-1.7359, -1.7359, -1.7359, -1.7359]],

        [[ 0.7218,  0.7218,  0.7218,  0.7218],
         [-1.7359, -1.7359, -1.7359, -1.7359]],

        [[ 0.7218,  0.7218,  0.7218,  0.7218],
         [-1.7359, -1.7359, -1.7359, -1.7359]]])
```

albanD replied:
What? Since when is expand doing broadcasting?? :p

zou3519 commented on tools/autograd/derivatives.yaml:
```diff
@@ -1202,18 +1223,23 @@
 
 - name: squeeze_(Tensor(a!) self) -> Tensor(a!)
   self: unsqueeze_to(grad, self.sizes())
+  result: self_t.squeeze()
```
Should this be in-place squeeze to match? (I'm not sure)

soulitzer replied:
Or in general for auto-linear inplace ops, can we avoid allocating the new self_t and modify the old self_t in place?

albanD replied:
Yes this should be linear so that it is inplace.

zou3519 commented on tools/autograd/derivatives.yaml:
```diff
@@ -1261,53 +1291,66 @@
 - name: triangular_solve(Tensor self, Tensor A, bool upper=True, bool transpose=False, bool unitriangular=False) -> (Tensor solution, Tensor cloned_coefficient)
   self, A: triangular_solve_backward(grads[0], grads[1], self, A, solution, upper, transpose, unitriangular, grad_input_mask)
 
 - name: tril(Tensor self, int diagonal=0) -> Tensor
   self: grad.tril(diagonal)
+  result: auto_linear
 
 - name: triu(Tensor self, int diagonal=0) -> Tensor
   self: grad.triu(diagonal)
+  result: auto_linear
```
Can these be auto_element_wise?

albanD replied:
They can, but the linear version is more efficient, so I prefer it when possible.

zou3519 commented on tools/autograd/derivatives.yaml:
```diff
@@ -1396,6 +1443,7 @@
 
 - name: zero_(Tensor(a!) self) -> Tensor(a!)
   self: zeros_like(grad)
+  result: auto_linear
```
Could this be auto_element_wise?

albanD replied:
No because element-wise does not handle the inplace case.

albanD replied:
We prefer this one to be inplace. So linear is prefered here.

zou3519 commented on tools/autograd/derivatives.yaml:
```diff
@@ -1537,22 +1585,23 @@
 
 - name: leaky_relu_(Tensor(a!) self, Scalar negative_slope=0.01) -> Tensor(a!)
   self: leaky_relu_backward(grad, result, negative_slope, true)
+  result: leaky_relu_backward(self_t, result, negative_slope, true)
```
What goes wrong if this is `auto_element_wise`?

albanD replied:
Nothing :D Changing.

zou3519 commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -6672,12 +6709,14 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
            dtypes=all_types_and_complex_and(torch.bool, torch.float16, torch.bfloat16),
            sample_inputs_func=sample_inputs_view_reshape,
            supports_out=False,
+           supports_forward_ad=True,
```
Does this opinfo only have contiguous inputs? If so, we should add some non-contiguous inputs to really check that reshape works.

https://github.com/pytorch/pytorch/blob/842a831f53bc9305c0553f3b2096afc6c1c6efe4/torch/testing/_internal/common_methods_invocations.py#L3954-L3968

albanD replied:
It wasn't indeed.
Updated,

zou3519 commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -5464,6 +5473,7 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
            # Batched grad checks fail for empty input tensors (see https://github.com/pytorch/pytorch/issues/53407)
            check_batched_grad=False,
            check_batched_gradgrad=False,
+           supports_forward_ad=True,
```
Random question but... what happens if someone tries to use a dual_tensor with an out= operation?

albanD replied:
Very good point. It silently ignores it for now I think.

albanD replied:
Actually it is only missing the check for the "out" argument. All the other ones are properly handled.

albanD replied:
Added a PR below this one to fix this.

zou3519 reviewed (commented):
I'm concerned about the test for reshape and the behavior of in-place views, but other than that this lgtm

zou3519 commented on tools/autograd/derivatives.yaml:
```diff
@@ -959,9 +959,11 @@
 
 - name: mvlgamma(Tensor self, int p) -> Tensor
   self: mvlgamma_backward(grad, self, p)
+  result: auto_element_wise
```
Btw, here's a list of all the ops that appear in this PR: https://docs.google.com/spreadsheets/d/16DgjrIyzLbNMOIRNF_nV7qxTFqBFapRwisoNftip2s0/edit?usp=sharing

albanD committed:
Update on "All remaining linear/element-wise formulas"




[ghstack-poisoned]

albanD committed:
Update on "All remaining linear/element-wise formulas"




[ghstack-poisoned]

albanD committed:
Update on "All remaining linear/element-wise formulas"




[ghstack-poisoned]

albanD committed:
Update on "All remaining linear/element-wise formulas"




[ghstack-poisoned]

albanD committed:
Update on "All remaining linear/element-wise formulas"




[ghstack-poisoned]

albanD committed:
Update on "All remaining linear/element-wise formulas"




[ghstack-poisoned]

albanD committed:
Update on "All remaining linear/element-wise formulas"




[ghstack-poisoned]

zou3519 commented on tools/autograd/derivatives.yaml:
```diff
@@ -1114,21 +1120,27 @@
 
 - name: remainder.Tensor(Tensor self, Tensor other) -> Tensor
   self: grad
+  # Use addition with zeros to get the proper broadcasting behavior
```
Comment is outdated

albanD committed:
Update on "All remaining linear/element-wise formulas"




[ghstack-poisoned]

albanD committed:
Update on "All remaining linear/element-wise formulas"




[ghstack-poisoned]

albanD committed:
Update on "All remaining linear/element-wise formulas"




[ghstack-poisoned]

albanD commented:
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29914594).

albanD committed:
Update on "All remaining linear/element-wise formulas"


Differential Revision: [D29914594](https://our.internmc.facebook.com/intern/diff/D29914594)

[ghstack-poisoned]

albanD commented:
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29914594).
